File: src/main/java/com/cxin/chronicle/controller/EntriesController.java
package com.cxin.chronicle.controller;

import com.cxin.chronicle.infrastructure.annotation.AuthCheck;
import com.cxin.chronicle.infrastructure.common.BaseResponse;
import com.cxin.chronicle.infrastructure.common.ResultUtils;
import com.cxin.chronicle.infrastructure.constant.UserConstant;
import com.cxin.chronicle.infrastructure.model.dto.entries.*;
import com.cxin.chronicle.infrastructure.model.entity.User;
import com.cxin.chronicle.infrastructure.model.vo.EntriesVo;
import com.cxin.chronicle.infrastructure.model.vo.HeatmapDataVo;
import com.cxin.chronicle.service.EntriesService;
import com.cxin.chronicle.service.UserService;
import io.swagger.v3.oas.annotations.Operation;
import jakarta.annotation.Resource;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * <p>
 * 记录表 前端控制器
 * </p>
 *
 * @author Charles Chen
 * @since 2026-04-25
 */
@RestController
@RequestMapping("/entries")
public class EntriesController {

    @Resource
    private EntriesService entriesService;
    @Resource
    private UserService userService;

    @PostMapping("/add")
    @Operation(summary = "添加记录")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    public BaseResponse<Boolean> createArticle(@RequestBody @Valid EntriesAddReq request, HttpServletRequest httpServletRequest) {
        User loginUser = userService.getCurrentUser(httpServletRequest);
        Boolean added = entriesService.addEntries(loginUser, request);
        return ResultUtils.success(added);
    }

    @PostMapping("/query/daily")
    @Operation(summary = "查询日记录")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    public BaseResponse<List<EntriesVo>> queryDailyEntries(@RequestBody @Valid EntriesQueryReq request, HttpServletRequest httpServletRequest) {
        User loginUser = userService.getCurrentUser(httpServletRequest);
        List<EntriesVo> entriesVoList = entriesService.queryDailyEntries(loginUser, request);
        return ResultUtils.success(entriesVoList);
    }

    @PostMapping("/query/heatmap")
    @Operation(summary = "查询热力图数据")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    public BaseResponse<List<HeatmapDataVo>> queryHeatmapData(
            @RequestParam String year,
            HttpServletRequest httpServletRequest) {
        User loginUser = userService.getCurrentUser(httpServletRequest);
        // 解析年份参数，格式：yyyy
        int yearInt = Integer.parseInt(year);
        List<HeatmapDataVo> heatmapDataList = entriesService.queryHeatmapData(loginUser, yearInt);
        return ResultUtils.success(heatmapDataList);
    }

    @PostMapping("/update/checked")
    @Operation(summary = "更新记录勾选状态")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    public BaseResponse<Boolean> updateChecked(@RequestBody @Valid EntriesUpdateCheckedReq request, HttpServletRequest httpServletRequest) {
        User loginUser = userService.getCurrentUser(httpServletRequest);
        Boolean updated = entriesService.updateChecked(loginUser, request);
        return ResultUtils.success(updated);
    }

    @PostMapping("/update/content-and-type")
    @Operation(summary = "更新记录内容和类型")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    public BaseResponse<Boolean> updateContentAndType(@RequestBody @Valid EntriesUpdateContentAndTypeReq request, HttpServletRequest httpServletRequest) {
        User loginUser = userService.getCurrentUser(httpServletRequest);
        Boolean updated = entriesService.updateContentAndType(loginUser, request);
        return ResultUtils.success(updated);
    }

    @PostMapping("/delete")
    @Operation(summary = "删除记录")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    public BaseResponse<Boolean> deleteEntry(@RequestBody @Valid EntriesDeleteReq request, HttpServletRequest httpServletRequest) {
        User loginUser = userService.getCurrentUser(httpServletRequest);
        Boolean deleted = entriesService.deleteEntry(loginUser, request);
        return ResultUtils.success(deleted);
    }
}

File: chronicle-frontend/src/api/entriesController.ts
// @ts-ignore
/* eslint-disable */
import request from '@/request'

/** 添加记录 POST /entries/add */
export async function addEntry(body: API.EntriesAddReq, options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/entries/add', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 删除记录 POST /entries/delete */
export async function deleteEntry(body: API.EntriesDeleteReq, options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/entries/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 查询日记录 POST /entries/query/daily */
export async function queryDailyEntries(
  body: API.EntriesQueryReq,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseListEntriesVo>('/entries/query/daily', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 查询热力图数据 POST /entries/query/heatmap */
export async function queryHeatmapData(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.queryHeatmapDataParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseListHeatmapDataVo>('/entries/query/heatmap', {
    method: 'POST',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}

/** 游标分页查询历史记录 POST /entries/query/history */
export async function queryHistoryEntries(
  body: API.EntriesHistoryReq,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseListEntriesVo>('/entries/query/history', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 更新记录勾选状态 POST /entries/update/checked */
export async function updateChecked(
  body: API.EntriesUpdateCheckedReq,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>('/entries/update/checked', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 更新记录（内容、类型、完成百分比） POST /entries/update/content */
export async function updateEntriesContent(
  body: API.EntriesContentUpdateReq,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>('/entries/update/content', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}
